src/function/maybe_changed_after.rs

```diff
@@ -433,14 +456,17 @@ where
                         );
 
                         if in_heads {
-                            // Iterate our dependency graph again, starting from the top. We clear the
-                            // cycle heads here because we are starting a fresh traversal. (It might be
-                            // logically clearer to create a new HashSet each time, but clearing the
-                            // existing one is more efficient.)
-                            cycle_heads.clear();
+                            if is_provisional {
+                                old_memo
```

Comment:
I don't think this is technically required because a later read will set the `finalize` flag but doing it here allows short-circuting in later reads (e.g. adding an `assert` here that `verified_final` is `true` panics.